### ISSUE TITLE
Missing support for project reference

### ISSUE THREAD

**Issue by armano2** [typescript-eslint/typescript-eslint#1545]
I'm was testing new setup within this project and i noticed that project references are not supported

https://www.typescriptlang.org/docs/handbook/project-references.html

```js
// https://github.com/typescript-eslint/typescript-eslint/blob/master/.eslintrc.js
module.exports = {
  root: true,
  plugins: [
    '@typescript-eslint'
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
  // ....
  parserOptions: {
    sourceType: 'module',
    project: './tsconfig.json',
    tsconfigRootDir: __dirname,
  },
}
```

```json5
// tsconfig.json
{
    "files": [],
    "include": [],
    "extends": "./tsconfig.base.json",
    "references": [
        { "path": "packages/typescript-estree" },
        { "path": "packages/parser" },
        { "path": "packages/experimental-utils" },
        { "path": "packages/eslint-plugin" },
        { "path": "packages/eslint-plugin-tslint" },
        { "path": "packages/eslint-plugin-internal" }
   ]
}
```

each of packages has tsconfig with `composite: true`

**What did you expect to happen?**
Persisted parse should correctly resolve references and lint files.

**What actually happened?**
Persisted parse was unable to resolve references and file is not in project error appeared.

```
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: *ANY_FILE*.
The file must be included in at least one of the projects provided
```

**Versions**

| package                                               | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `master` |
| `TypeScript`                                          | `3.7.4` |
| `node`                                                | `unrelated` |
| `npm`                                                | `unrelated` |

**Comment by bradzacher** [comment]
We don't do any tsconfig parsing, so we rely upon typescript's mechanisms.
In particular, `createWatchProgram` doesn't support project references, so we don't either.

See discussion in #856, in particular: https://github.com/typescript-eslint/typescript-eslint/issues/856#issuecomment-521729080

>> Is TS supposed to "include" all of the files in the referenced projects, or are they consumed via a different mechanism?
>
> They are consumed by a different mechanism - in particular, the goal of project references is that when project a references project b, project a does not include the .ts files in project b.

**Comment by armano2** [comment]
i did some debugging and `createWatchProgram` is partially supporting references, its reading directories and resolves paths correctly

`getProjectReferences` is filled correctly same goes for `getResolvedProjectReferences`
